wendy.1
.Dd 2013-06-02
.Dt WENDY 1
.Os POSIX.1-2008
.Sh NAME
.Nm wendy
.Nd inotify based event watcher
.Sh SYNOPSIS
.Nm wendy
.Op Fl adlrv
.Op Fl m Ar mask
.Op Fl w Ar inode
.Op command Op Ar args...
.Sh DESCRIPTION
.Nm
watches for events in a directory or its files and executes
.Ar cmd
when an inotify event is triggered.
.Bl -tag -width Ds
.It Fl a
Hidden mode. Entries whose name start with a . will be listed and trigger
an event.
.It Fl d
Directory mode. Only directories will be watched, while regular
files will be skipped (even if explicitely set with -w). See IN_ONLYDIR
from
.Xr inotify 7 .
.It Fl l
List events along with their numeric values.
.It Fl r
Recursive mode. Everytime an IN_CREATE event is triggered,
a watch is added on the target file/directory.
.It Fl v
Verbose mode. Each triggered event is printed along with the path,
as provided on the command line. In the case of directories, the path
triggering the event is concatenated to the directory path.
.It Fl m Ar mask
Sets the mask for events you are interested in (see MASKS for more details).
default: IN_CREATE|IN_DELETE|IN_DELETE_SELF|IN_MODIFY|IN_MOVE|IN_MOVE_SELF|IN_CLOSE_WRITE
.It Fl w Ar inode
Specifies the file or directory to watch events in. You can specify multiple
files by providing the
.Fl w
flag multiple times. If no watcher is added with the -w flag,
.Nm
will read path names from
.Ar stdin ,
one per line.
.El
.Sh MASKS
The mask is a numeric value passed to inotify to specify the events that should
be raised. Multiple events can be registered by summing them up, and passing
the result to the
.Fl m
flag. The mask can take any of the values below, or a sum of them:

.Bl -bullet -compact
.It
IN_ACCESS          1
.It
IN_MODIFY          2
.It
IN_ATTRIB          4
.It
IN_CLOSE_WRITE     8
.It
IN_CLOSE_NOWRITE   16
.It
IN_CLOSE           24
.It
IN_OPEN            32
.It
IN_MOVED_FROM      64
.It
IN_MOVED_TO        128
.It
IN_MOVE            192
.It
IN_CREATE          256
.It
IN_DELETE          512
.It
IN_DELETE_SELF     1024
.It
IN_MOVE_SELF       2048
.It
IN_ALL_EVENTS      4095
.It
IN_UNMOUNT         8192
.El
.Sh ENVIRONMENT
.Bl -tag -width Ds
.It Ev WENDY_INODE
Name of the inode concerned by the event
.It Ev WENDY_EVENT
Name of the event triggered (eg. "CREATE")
.El
.Sh EXAMPLES
Whenever an IN_CREATE (256) event is triggered in the Maildir folder
INBOX, a sound is played:
.Bd -literal
  wendy -m 256 -w ~/mail/INBOX/new play /usr/share/sound/bell.wav
.Ed
.Pp
Use a sub-shell to print event name and path to stdout, showcasing the
variables exported internally (this replicates the
.Fl v
switch):
.Bd -literal
  wendy -m 768 -w . sh -c 'echo $WENDY_EVENT $WENDY_INODE'
.Ed
.Pp
Rebuild a project when one of its source files changes:
.Bd -literal
  find . -name '*.[ch]' | wendy make
.Ed
.Pp
This example will only watch directory nodes, and will monitor every
new directory created, thus propagating event monitoring inside the tree.
Note that events happening on files withing these directories will
still be reported, making the following command suitable for
synchronizing two directories:
.Bd -literal
  find /src -type d | wendy -rd -m 970 rsync -av --delete /src /dest
.Ed
.Sh SEE ALSO
.Xr inotify 7 ,
.Xr rsync 1
.Sh AUTHORS
.An Willy Goiffon Aq Mt dev@z3bra.org


wendy.c
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <limits.h>
#include <string.h>
#include <sys/inotify.h>
#include <sys/wait.h>

#include "arg.h"
#include "queue.h"
#include "strlcpy.h"

#define EVSZ (sizeof(struct inotify_event) + NAME_MAX + 1)
#define MASK (IN_CREATE|IN_DELETE|IN_DELETE_SELF|IN_MODIFY|IN_MOVE|IN_MOVE_SELF|IN_CLOSE_WRITE)

struct watcher {
	int wd;
	char path[PATH_MAX];
	SLIST_ENTRY(watcher) entries;
};

SLIST_HEAD(watchers, watcher) head;

char *evname[IN_ALL_EVENTS] = {
	[IN_ACCESS] =        "ACCESS",
	[IN_MODIFY] =        "MODIFY",
	[IN_ATTRIB] =        "ATTRIB",
	[IN_CLOSE_WRITE] =   "CLOSE_WRITE",
	[IN_CLOSE_NOWRITE] = "CLOSE_NOWRITE",
	[IN_OPEN] =          "OPEN",
	[IN_MOVED_FROM] =    "MOVED_FROM",
	[IN_MOVED_TO] =      "MOVED_TO",
	[IN_CREATE] =        "CREATE",
	[IN_DELETE] =        "DELETE",
	[IN_DELETE_SELF] =   "DELETE_SELF",
	[IN_MOVE_SELF] =     "MOVE_SELF",
};

int verbose = 0;
int aflag = 0, dflag = 0, lflag = 0, rflag = 0;

void
usage(char *name)
{
	fprintf(stderr, "usage: %s [-adrv] [-m mask] [-w file] [command [args...]]\n", name);
	exit(1);
}

char *
basename(char *p)
{
	char *b = strrchr(p, '/');
	return b ? b + 1 : p;
}

int
listevents(char **ev)
{
	int i;
	for (i=0; i<IN_ALL_EVENTS; i++)
		if (ev[i])
			printf("%s\t%d\n", ev[i], i);
	return 0;
}

struct watcher *
getwatcher(struct watchers *h, int wd)
{
	struct watcher *tmp;

	SLIST_FOREACH(tmp, h, entries)
		if (tmp->wd == wd)
			return tmp;

	return NULL;
}

struct watcher *
watch(int fd, char *pathname, int mask)
{
	size_t len;
	struct watcher *w;

	w = malloc(sizeof(*w));
	if (!w)
		return NULL;

	/* store inode path, eventually removing trailing '/' */
	len = strlcpy(w->path, pathname, PATH_MAX);
	if (w->path[len - 1] == '/')
		w->path[len - 1] = '\0';

	w->wd = inotify_add_watch(fd, w->path, mask);
	if (w->wd < 0) {
		/* triggered when dflag is set, so it is expected */
		if (errno != ENOTDIR)
			perror(pathname);

		free(w);
		return NULL;
	}

	SLIST_INSERT_HEAD(&head, w, entries);
	return w;
}

int
watchstream(int fd, FILE *stream, int mask)
{
	size_t l, n = 0;
	char *p = NULL;

	while (getline(&p, &n, stream) > 0) {
		l = strlen(p);
		p[l-1] = '\0';
		watch(fd, p, mask);
	}
	free(p);

	return 0;
}

char *
wdpath(struct inotify_event *e, struct watcher *w)
{
	size_t len;
	static char pathname[PATH_MAX];

	len = strlcpy(pathname, w->path, PATH_MAX);
	if (e->len) {
		strncat(pathname, "/", PATH_MAX - len - 1);
		len = strnlen(pathname, PATH_MAX - 1);
		strncat(pathname, e->name, PATH_MAX - len - 1);
	}

	return pathname;
}


int
main (int argc, char **argv)
{
	int fd;
	uint8_t buf[EVSZ];
	uint32_t mask = MASK;
	ssize_t len, off = 0;
	char **cmd, *argv0 = NULL;
	struct watcher *w;
	struct inotify_event *e;

	/* get file descriptor */
	fd = inotify_init();
	if (fd < 0)
		perror("inotify_init");

	ARGBEGIN {
	case 'a':
		aflag = 1;
		break;
	case 'd':
		dflag = 1;
		break;
	case 'l':
		lflag = 1;
		break;
	case 'r':
		rflag = 1;
		break;
	case 'm':
		mask = atoi(EARGF(usage(argv0)));
		break;
	case 'v':
		verbose++;
		break;
	case 'w':
		watch(fd, EARGF(usage(argv0)), mask);
		break;
	default:
		usage(argv0);
	} ARGEND;

	if (lflag) {
		listevents(evname);
		return 0;
	}

	/* remaining arguments is the command to run on event */
	cmd = (argc > 0) ? argv : NULL;

	/* ensure that only directories are watched for */
	if (dflag)
		mask |= IN_ONLYDIR;

	if (SLIST_EMPTY(&head))
		watchstream(fd, stdin, mask);

	while (!SLIST_EMPTY(&head) && (off || (len=read(fd, buf, EVSZ))>0)) {

		/* cast buffer into the event structure */
		e = (struct inotify_event *) (buf + off);

		/* skip watch descriptors not in out list */
		if (!(w = getwatcher(&head, e->wd))) {
			inotify_rm_watch(fd, e->wd);
			goto skip;
		}

		/* skip hidden files when aflag is not set */
		if (!aflag && basename(wdpath(e, w))[0] == '.')
			goto skip;

		if (verbose && e->mask & IN_ALL_EVENTS) {
			printf("%s\t%s\n", evname[e->mask & IN_ALL_EVENTS], wdpath(e, w));
			fflush(stdout);
		}

		/* orphan process */
		if (cmd) {
			if (!fork()) {
				if (!fork()) {
					setenv("WENDY_INODE", wdpath(e, w), 1);
					setenv("WENDY_EVENT", evname[e->mask & IN_ALL_EVENTS], 1);
					execvp(cmd[0], cmd);
					return -1; /* NOTREACHED */
				}
				return 0;
			}
			wait(NULL);
		}

		switch(e->mask & (IN_ALL_EVENTS|IN_IGNORED)) {
		case IN_CREATE:
			/* Watch subdirectories upon creation */
			if (rflag)
				watch(fd, wdpath(e, w), mask);
			break;

		/*
		 * IN_IGNORED is triggered when a file watched
		 * doesn't exists anymore. In this case we first try to
		 * watch it again (in case it was recreated), and if it
		 * fails, remove the watcher completely.
		 */
		case IN_IGNORED:
			inotify_rm_watch(fd, e->wd);
			if ((w->wd = inotify_add_watch(fd, w->path, mask)) < 0) {
				SLIST_REMOVE(&head, w, watcher, entries);
				free(w);
			}
			break;
		}

skip:
		/* shift buffer offset when there's more events to read */
		off += sizeof(*e) + e->len;
		if (off >= len)
			off = 0;
	}

	close(fd);

	return 0;
}
